File: sdk-python/src/personaai/resources/agents.py
"""Agents (``/api/v1/developer/agents``) — Project-owned, or, when this
client asserts an external user, owned by that end user. Ported from
``sdk/src/resources/agents.ts``."""

from __future__ import annotations

from typing import TYPE_CHECKING, List, cast

from ..types.agent import Agent, CreateAgentInput, DiscoverAgentsParams, UpdateAgentInput
from ..types.bulk_delete import BulkDeleteResult
from ..types.pagination import PaginatedResult

if TYPE_CHECKING:
    from .._async_http import AsyncTransport
    from .._sync_http import SyncTransport


class Agents:
    def __init__(self, transport: SyncTransport) -> None:
        self._transport = transport

    def create(self, input: CreateAgentInput, idempotency_key: str | None = None) -> Agent:
        """``idempotency_key``, if provided, is sent as the ``Idempotency-Key``
        header — a safe retry with the same key replays the original
        response instead of creating a duplicate Agent."""
        headers = {"Idempotency-Key": idempotency_key} if idempotency_key else None
        return cast(
            Agent,
            self._transport.request(
                "POST", "/api/v1/developer/agents", json=input, headers=headers
            ),
        )

    def list(self, params: DiscoverAgentsParams | None = None) -> PaginatedResult[Agent]:
        return cast(
            PaginatedResult[Agent],
            self._transport.request("GET", "/api/v1/developer/agents", query=params),
        )

    def get(self, agent_id: str) -> Agent:
        """Populates ``skills``/``mcps``/``knowledgeBases`` as objects
        (unlike ``create``/``update``/``list``)."""
        return cast(Agent, self._transport.request("GET", f"/api/v1/developer/agents/{agent_id}"))

    def update(self, agent_id: str, input: UpdateAgentInput) -> Agent:
        return cast(
            Agent,
            self._transport.request("PATCH", f"/api/v1/developer/agents/{agent_id}", json=input),
        )

    def delete(self, agent_id: str) -> None:
        self._transport.request("DELETE", f"/api/v1/developer/agents/{agent_id}")

    def bulk_delete(self, ids: List[str]) -> BulkDeleteResult:
        """Best-effort batch delete — up to 100 ids per call; partial failures don't raise."""
        return cast(
            BulkDeleteResult,
            self._transport.request(
                "POST", "/api/v1/developer/agents/bulk-delete", json={"ids": ids}
            ),
        )


class AsyncAgents:
    def __init__(self, transport: AsyncTransport) -> None:
        self._transport = transport

    async def create(self, input: CreateAgentInput, idempotency_key: str | None = None) -> Agent:
        """``idempotency_key``, if provided, is sent as the ``Idempotency-Key``
        header — a safe retry with the same key replays the original
        response instead of creating a duplicate Agent."""
        headers = {"Idempotency-Key": idempotency_key} if idempotency_key else None
        return cast(
            Agent,
            await self._transport.request(
                "POST", "/api/v1/developer/agents", json=input, headers=headers
            ),
        )

    async def list(self, params: DiscoverAgentsParams | None = None) -> PaginatedResult[Agent]:
        return cast(
            PaginatedResult[Agent],
            await self._transport.request("GET", "/api/v1/developer/agents", query=params),
        )

    async def get(self, agent_id: str) -> Agent:
        """Populates ``skills``/``mcps``/``knowledgeBases`` as objects
        (unlike ``create``/``update``/``list``)."""
        return cast(
            Agent, await self._transport.request("GET", f"/api/v1/developer/agents/{agent_id}")
        )

    async def update(self, agent_id: str, input: UpdateAgentInput) -> Agent:
        return cast(
            Agent,
            await self._transport.request(
                "PATCH", f"/api/v1/developer/agents/{agent_id}", json=input
            ),
        )

    async def delete(self, agent_id: str) -> None:
        await self._transport.request("DELETE", f"/api/v1/developer/agents/{agent_id}")

    async def bulk_delete(self, ids: List[str]) -> BulkDeleteResult:
        """Best-effort batch delete — up to 100 ids per call; partial failures don't raise."""
        return cast(
            BulkDeleteResult,
            await self._transport.request(
                "POST", "/api/v1/developer/agents/bulk-delete", json={"ids": ids}
            ),
        )

File: sdk/typescript/src/resources/agents.ts
import type { HttpClient } from '../http.js';
import type {
  Agent,
  CreateAgentInput,
  DiscoverAgentsParams,
  McpConnection,
  UpdateAgentInput,
} from '../types/agent.js';
import type { BulkDeleteResult } from '../types/bulkDelete.js';
import type { PaginatedResult } from '../types/pagination.js';

/**
 * Agents (`/api/v1/developer/agents`) — Project-owned, or, when this client
 * asserts an external user, owned by that end user.
 */
export class AgentsResource {
  constructor(private readonly http: HttpClient) {}

  /**
   * Creates a new Agent.
   * @param input - `name`, `systemPrompt`, `providerId` are required.
   * @param idempotencyKey - Optional. Sent as the `Idempotency-Key` header —
   *   a safe retry with the same key replays the original response instead
   *   of creating a duplicate Agent.
   * @returns The created {@link Agent} (raw Mongo shape — `_id`, not `id`).
   */
  async create(input: CreateAgentInput, idempotencyKey?: string): Promise<Agent> {
    return this.http.request<Agent>('POST', '/api/v1/developer/agents', {
      body: input,
      headers: idempotencyKey ? { 'Idempotency-Key': idempotencyKey } : undefined,
    });
  }

  /**
   * Lists/searches Agents visible to this credential (this Project's own,
   * plus any public Agents).
   * @param params - `page` (default `1`), `limit` (default `20`), `search`
   *   (free-text), `category`, `scope: 'mine'` (restricts to the asserted
   *   external user's own Agents — `ProjectRuntimeContext` only).
   * @returns `{ items, pagination: { total, page, limit, pages } }`. Note:
   *   `skills`/`mcps`/`knowledgeBases` on each item are bare id strings here
   *   (unlike `get()`, which populates them as objects).
   */
  async list(params: DiscoverAgentsParams = {}): Promise<PaginatedResult<Agent>> {
    return this.http.request<PaginatedResult<Agent>>('GET', '/api/v1/developer/agents', {
      query: { ...params },
    });
  }

  /**
   * Fetches a single Agent by id.
   * @param agentId - The Agent's `_id`.
   * @returns The {@link Agent}, with `skills`/`mcps`/`knowledgeBases`
   *   populated as objects (unlike `create`/`update`/`list`).
   */
  async get(agentId: string): Promise<Agent> {
    return this.http.request<Agent>('GET', `/api/v1/developer/agents/${agentId}`);
  }

  /**
   * Partially updates an Agent — only the fields you pass are changed.
   * @param agentId - The Agent's `_id`.
   * @param input - Any subset of the fields on {@link Agent} (excluding server-managed ones).
   */
  async update(agentId: string, input: UpdateAgentInput): Promise<Agent> {
    return this.http.request<Agent>('PATCH', `/api/v1/developer/agents/${agentId}`, {
      body: input,
    });
  }

  /**
   * Deletes an Agent.
   * @param agentId - The Agent's `_id`.
   */
  async delete(agentId: string): Promise<void> {
    await this.http.request<unknown>('DELETE', `/api/v1/developer/agents/${agentId}`);
  }

  /**
   * Best-effort batch delete — partial failures don't throw or abort the
   * rest of the batch.
   * @param ids - Up to 100 Agent ids per call.
   * @returns `{ deleted, failed }` — check `failed` for per-id reasons.
   */
  async bulkDelete(ids: string[]): Promise<BulkDeleteResult> {
    return this.http.request<BulkDeleteResult>('POST', '/api/v1/developer/agents/bulk-delete', {
      body: { ids },
    });
  }

  /**
   * Per-user OAuth connection status for every `authType: 'oauth',
   * authMode: 'user'` MCP this Agent has attached. Requires the client to
   * assert an external user (`ProjectRuntimeContext`) — there's no
   * per-user connection concept for a bare Project credential.
   * @param agentId - The Agent's `_id`.
   * @param returnTo - Where to send the browser after OAuth consent
   *   completes, for any MCP that isn't connected yet. Defaults to this
   *   platform's own site if omitted.
   */
  async getMcpConnections(agentId: string, returnTo?: string): Promise<McpConnection[]> {
    return this.http.request<McpConnection[]>(
      'GET',
      `/api/v1/developer/agents/${agentId}/mcp-connections`,
      { query: returnTo ? { returnTo } : undefined }
    );
  }
}
